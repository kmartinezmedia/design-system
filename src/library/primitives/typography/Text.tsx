// @CB Text Base Component - should not be imported outside of design system
import React from 'react';
import {
  Text as RNText,
  TextStyle,
  TextProps as RNTextProps,
  StyleSheet,
} from 'react-native';
import { SurfaceColorMap, Spacing } from '@designSystem/types';
import { debounce } from '@utils/Events';
import { useSpacing, useForeground } from '@designSystem/hooks';

export interface TextProps<
  T extends keyof SurfaceColorMap = keyof SurfaceColorMap,
  C extends SurfaceColorMap[T] = SurfaceColorMap[T]
> extends RNTextProps {
  align?: TextStyle['textAlign'];
  ellipsize?: RNTextProps['ellipsizeMode'];
  style?: TextStyle;
  spacing?: Spacing;
  lineHeight?: number | 'none';
  dangerouslySetStyle?: TextStyle;
  surface?: T;
  color?: C;
}

export const Text = <
  T extends keyof SurfaceColorMap,
  C extends SurfaceColorMap[T]
>({
  surface = 'background' as T,
  color = 'default' as C,
  children,
  align,
  ellipsize,
  style,
  onPress,
  spacing,
  lineHeight,
  dangerouslySetStyle,
  ...textProps
}: React.PropsWithChildren<TextProps<T, C>>) => {
  const textColor = useForeground(surface, color);

  const computedLineHeight =
    lineHeight === undefined ? style?.lineHeight : lineHeight;

  const ellipsizeProps =
    ellipsize &&
    (() => {
      return {
        numberOfLines: textProps.numberOfLines || 1,
        ellipsizeMode: ellipsize,
      };
    })();

  return (
    <RNText
      onPress={onPress && debounce(onPress)}
      style={[
        style,
        {
          color: textColor,
          textAlign: align,
          lineHeight:
            computedLineHeight === 'none' ? undefined : computedLineHeight,
        },
        ellipsize && styles.withEllipse,
        useSpacing(spacing),
        dangerouslySetStyle,
      ]}
      {...ellipsizeProps}
      {...textProps}
    >
      {children}
    </RNText>
  );
};

const styles = StyleSheet.create({
  withEllipse: {
    overflow: 'hidden',
  },
});
