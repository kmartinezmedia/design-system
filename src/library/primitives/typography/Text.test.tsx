import React from 'react';
import { render, toJSON } from '@testing-library/react-native';

import { Text } from './Text';

const textColors = [
  'default',
  'muted',
  'primary',
  'negative',
  'positive',
  'warning',
] as const;
const align = ['left', 'center', 'right'] as const;
const ellipsize = ['clip', 'head', 'middle', 'tail'] as const;

const TestComponent = () => (
  <>
    <Text>Default</Text>
    <Text spacing={{ top: 2, horizontal: 3 }}>Spacing</Text>
    <Text lineHeight={1}>Line Height</Text>
    {textColors.map((color) => (
      <Text key={color} color={color}>
        {color}
      </Text>
    ))}
    {align.map((direction) => (
      <Text key={direction} align={direction}>
        {direction}
      </Text>
    ))}
    {ellipsize.map((type) => (
      <Text key={type} ellipsize={type}>
        {type}
      </Text>
    ))}
  </>
);

describe('Text', () => {
  async function renderSnapshot(lookupText: string) {
    const { getByText, rerender } = await render(<TestComponent />);
    const el = getByText(lookupText);
    expect(toJSON(el)).toMatchSnapshot();
    return { el, rerender };
  }

  it.only('default', async () => {
    const { el, rerender } = await renderSnapshot('Default');
    const style1 = el.getProp('style');
    await rerender();
    expect(el.getProp('style')).toEqual(style1);
  });
  it('spacing', async () => {
    await renderSnapshot('Spacing');
  });
  it('color', async () => {
    await Promise.all(textColors.map((color) => renderSnapshot(color)));
  });
  it('align', async () => {
    await Promise.all(align.map((direction) => renderSnapshot(direction)));
  });
  it('ellipsize', async () => {
    await Promise.all(ellipsize.map((type) => renderSnapshot(type)));
  });
  it('line height', async () => {
    await renderSnapshot('Line Height');
  });
});
