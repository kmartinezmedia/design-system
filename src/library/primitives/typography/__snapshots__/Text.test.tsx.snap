// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text align 1`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": undefined,
        "textAlign": "left",
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  left
</Text>
`;

exports[`Text align 2`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": undefined,
        "textAlign": "center",
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  center
</Text>
`;

exports[`Text align 3`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": undefined,
        "textAlign": "right",
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  right
</Text>
`;

exports[`Text color 1`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  default
</Text>
`;

exports[`Text color 2`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#587088",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  muted
</Text>
`;

exports[`Text color 3`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#1652F0",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  primary
</Text>
`;

exports[`Text color 4`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#E5212E",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  negative
</Text>
`;

exports[`Text color 5`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#00A45F",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  positive
</Text>
`;

exports[`Text color 6`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#FF9900",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  warning
</Text>
`;

exports[`Text default 1`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  Default
</Text>
`;

exports[`Text ellipsize 1`] = `
<Text
  ellipsizeMode="clip"
  numberOfLines={1}
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      Object {
        "overflow": "hidden",
      },
      Object {},
      undefined,
    ]
  }
>
  clip
</Text>
`;

exports[`Text ellipsize 2`] = `
<Text
  ellipsizeMode="head"
  numberOfLines={1}
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      Object {
        "overflow": "hidden",
      },
      Object {},
      undefined,
    ]
  }
>
  head
</Text>
`;

exports[`Text ellipsize 3`] = `
<Text
  ellipsizeMode="middle"
  numberOfLines={1}
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      Object {
        "overflow": "hidden",
      },
      Object {},
      undefined,
    ]
  }
>
  middle
</Text>
`;

exports[`Text ellipsize 4`] = `
<Text
  ellipsizeMode="tail"
  numberOfLines={1}
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      Object {
        "overflow": "hidden",
      },
      Object {},
      undefined,
    ]
  }
>
  tail
</Text>
`;

exports[`Text line height 1`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": 1,
        "textAlign": undefined,
      },
      undefined,
      Object {},
      undefined,
    ]
  }
>
  Line Height
</Text>
`;

exports[`Text spacing 1`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "color": "#071422",
        "lineHeight": undefined,
        "textAlign": undefined,
      },
      undefined,
      Object {
        "paddingHorizontal": 12,
        "paddingTop": 8,
      },
      undefined,
    ]
  }
>
  Spacing
</Text>
`;
